src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.P2P.Peer
+{
+    /// <summary>
+    /// Represents a TCP client that can connect to a server and send and receive messages.
+    /// </summary>
+    public class NetworkPeerClient : IDisposable
+    {
+        /// <summary>Logger factory to create loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private ILogger logger;
+
+        /// <summary>Unique identifier of a client.</summary>
+        public int Id { get; private set; }
+
+        /// <summary>Underlaying TCP client.</summary>
+        private TcpClient tcpClient;
+
+        /// <summary>Prevents parallel execution of multiple write operations on <see cref="Stream"/>.</summary>
+        private AsyncLock writeLock;
+
+        /// <summary>Stream to send and receive messages through established TCP connection.</summary>
+        /// <remarks>Write operations on the stream have to be protected by <see cref="writeLock"/>.</remarks>
+        public NetworkStream Stream { get; private set; }
+
+        /// <summary>Task that cares about the client once its connection is accepted up until the communication is terminated.</summary>
+        /// <remarks>The client is being processed as long as this task is not complete.</remarks>
+        public TaskCompletionSource<bool> ProcessingCompletion { get; private set; }
+
+        /// <summary><c>1</c> if the instance of the object has been disposed or disposing is in progress, <c>0</c> otherwise.</summary>
+        private int disposed;
```

Comment:
Can we use a bool instead here? IsDisposing - pretty sure this has already been mentioned and coming in part xx :grin:

Reply:
we can't, https://msdn.microsoft.com/en-us/library/system.threading.interlocked%28v=vs.110%29.aspx - Interlocked.CompareExchange is not available for bool types, we are currently using this for lockless synchronization, but maybe in a future PR we won't need that sync, I'm not sure yet, but quick answer is no

Reply:
Ok sounds good.